storage/remote/queue_manager.go

```diff
@@ -32,13 +33,25 @@ const (
 	subsystem = "remote_storage"
 	queue     = "queue"
 
-	defaultShards            = 10
+	// With a maximum of 500 shards, assuming an average of 100ms remote write
+	// time and 100 samples per batch, we will be able to push 500k samples/s.
+	defaultMaxShards         = 500
 	defaultMaxSamplesPerSend = 100
-	// The queue capacity is per shard.
-	defaultQueueCapacity     = 100 * 1024 / defaultShards
+
+	// defaultQueueCapacity is per shard - at 500 shards, this will buffer
+	// 50m samples.  It is configured to buffer 1024 batches, which at 100ms
+	// per batch is 1:40mins.
+	defaultQueueCapacity     = defaultMaxSamplesPerSend * 1024
 	defaultBatchSendDeadline = 5 * time.Second
-	logRateLimit             = 0.1 // Limit to 1 log event every 10s
-	logBurst                 = 10
+
+	// We track samples in/out and how long pushes take using an Exponentially
+	// Weighted Moving Average.
+	ewmaWeight             = 0.2
+	shardUpdateDuration    = 10 * time.Second
+	shardToleranceFraction = 0.3 // allow 30% too many shards before scaling down
```

Comment:
Have you considered keeping the number of shards always a power of 2? That may give the other side more options in terms of their own processing, as it'll tend to keep certain timeseries together through reshards.